src/store/base.py

```diff
@@ -0,0 +1,56 @@
+from abc import ABC, abstractmethod
+import os
+from SPARQLWrapper import SPARQLWrapper, JSON, QueryResult
+
+class BaseStore(ABC):
+
+    # Every client runs the setup method
+    def __init__(self):
+        self.setup()
+
+    # Every client must have a setup method
+    @abstractmethod
+    def setup(self):
+        ...
+
+    @abstractmethod
+    def get_datasets(self):
+        ...
+
+
+class SparqlStore(BaseStore):
+
+    #seting up the self.url
+    def setup(self):
+        # set the url for sparql
+        self.url = os.getenv("SPARQL_ENDPOINT_URL")
+
+        #to make sure self.url is not none providin a default url
+        if self.url == None:
+            self.url = "https://beta.gss-data.org.uk/sparql"
+
+
+    def get_datasets(self) -> QueryResult:
```

Comment:
so there's a point on restructuring here and a point on return types.

### restructuring

- 1.) We want to defined the _query string_ in the get_datasets() function, nice ✅ 
- 2.) We want _everything that uses the query string_ elsewhere in `self.run_sparql()`, ready for when we have many queries to run.

as an example (I've added some extra functions to get the idea across:

```

    def run_sparql(self, query):
        """ Runs and returns the results from a sparql query"""
        sparql = SPARQLWrapper(self.url)

        sparql.setQuery(query)
        sparql.setReturnFormat(JSON)
        return sparql.query()

    def get_datasets(self):
        """
        Get many datasets
        """
        query = """the sparql query to get many datasets"""
        result =  self.run_sparql(query)
        
    def get_datasets(self):
        """
        Get a single datasets
        """
        query = """the sparql query for getting one dataset"""
        result = self.run_sparql(query)
        
     def get_editions(self):
        """
        Get many editions
        """
        query = """the sparql query to get many editions"""
        result = self.run_sparql(query)
        
    def get_edition(self):
        """
        Get a single edition
        """
        query = """the sparql query for getting one edition"""
        result = self.run_sparql(query) 
```

does that makes more sense? With the current pattern we'll end up repeating ourselves quite a lot.

### return types

cant be QueryResult 😄 . The idea here is that _all_ implementations of a store have the same methods, take the same arguments and return the same thing - in this instance QueryResult is SPARQL specific.

think of it this way, if your next task was to create a SqlStore as an option we could use instead of SparqlStore then SqlStore.get_datasets() wont ever be returning a SPARQLWrapper.QueryResult, the thing being returned _has to be generic_.

Some sort of sensible dictionary structure is fine for now (we'll be updating the query in due course I'm sure), but make sure it is a dictionary and not a package specific class please.

Reply:
Yes agree with everything Mike said here, my bad on not catching these/explaining when we chatted the other day